File: tp2/src/CrazyTaxi/Assets/Scripts/Player/CarControleScript.js
@script AddComponentMenu ("CarPhys/Scripts/Car Control Script")
#pragma strict
var centerOfMass : Vector3;	//Center of mass
var dataWheel : WheelCollider;	//Wheel Collider from which you want to calculate the speed
var lowestSteerAtSpeed : float = 50;	//if lowestSteerAtSpeed < currentSpeed the steer Angle = highSpeedSteerAngel
var lowSpeedSteerAngel : float = 10;	//This could be a high value
var highSpeedSteerAngel : float = 1;	//This shouldn't be a high value (recomended for stability of car)
var decellarationSpeed : float = 30;	//How fast the car will decellarate
var maxTorque : float  = 50;	//Maximum Torque
var currentSpeed : float;		//Current Speed of car
var topSpeed : float = 150;		//Highest speed at which the car can go
var maxReverseSpeed : float = 50; 	//Highest Reverse speed
var backLightObject : GameObject;	//Mesh for reverse light
var idleLightMaterial : Material;	//for idle state
var brakeLightMaterial : Material; 	//Braked state
var reverseLightMaterial : Material;	//Reverse state
@HideInInspector	
var braked : boolean = false;	//Brake trigger
var maxBrakeTorque : float = 100; 	//Braking speed
var speedOMeterDial : Texture2D;	//GUI Texture for dial
var speedOMeterPointer : Texture2D;		//GUI Texture for needle
var gearRatio : int[];		//Shift gear at speed
//var spark : GameObject;		//OnCollision Spark
var collisionSound : GameObject;	//OnCollision Sound
var minAnglePointer : int = -90;
var maxAnglePointer : int = 180;

function Start () {
rigidbody.centerOfMass=centerOfMass; //Center of mass , for this the car should be pointing on z axis
}

function FixedUpdate () {
HandBrake();
}
function Update(){
BackLight ();
EngineSound();
CalculateSpeed();
}

//Speed Calculation

function CalculateSpeed(){
currentSpeed = 2*22/7*dataWheel.radius*dataWheel.rpm*60/1000;
currentSpeed = Mathf.Round(currentSpeed);
}

//Light Control

function BackLight (){
if (currentSpeed > 0 && Input.GetAxis("Vertical")<0&&!braked){
backLightObject.renderer.material = brakeLightMaterial;
}
else if (currentSpeed < 0 && Input.GetAxis("Vertical")>0&&!braked){
backLightObject.renderer.material = brakeLightMaterial;
}
else if (currentSpeed < 0 && Input.GetAxis("Vertical")<0&&!braked){
backLightObject.renderer.material = reverseLightMaterial;
}
else if (!braked){
backLightObject.renderer.material = idleLightMaterial;
}
}

//Brake Trigger

function HandBrake(){
if (Input.GetButton("Jump")){
braked = true;
}
else{
braked = false;
}
}
//Engine Sound

function EngineSound(){
for (var i = 0; i < gearRatio.length; i++){
if(gearRatio[i]> currentSpeed){
break;
}
}
var gearMinValue : float = 0.00;
var gearMaxValue : float = 0.00;
if (i == 0){
gearMinValue = 0;
}
else {
gearMinValue = gearRatio[i-1];
}
gearMaxValue = gearRatio[i];
var enginePitch : float = ((currentSpeed - gearMinValue)/(gearMaxValue - gearMinValue))+1;
audio.pitch = enginePitch;
}

//Speedometer

function OnGUI (){
GUI.DrawTexture(Rect(Screen.width - 300,Screen.height - 300,300,300),speedOMeterDial);
var speedFactor : float = currentSpeed / topSpeed;
var rotationAngle : float;
if (currentSpeed >= 0){
  rotationAngle = Mathf.Lerp(minAnglePointer,maxAnglePointer,speedFactor);
  }
  else {
  rotationAngle = Mathf.Lerp(minAnglePointer,maxAnglePointer,-speedFactor);
  }
GUIUtility.RotateAroundPivot(rotationAngle,Vector2(Screen.width - 150 ,Screen.height - 150));
GUI.DrawTexture(Rect(Screen.width - 300,Screen.height - 300,300,300),speedOMeterPointer);



}

//CollisioN FX

function OnCollisionEnter (other : Collision){

if (other.transform != transform && other.contacts.length != 0){
for (var i = 0; i < other.contacts.length ; i++){
//Instantiate(spark,other.contacts[i].point,Quaternion.identity);
var clone : GameObject = Instantiate(collisionSound,other.contacts[i].point,Quaternion.identity);
clone.transform.parent = transform;
}
}
}

function OnDrawGizmos  () {
        Gizmos.color = Color.white;
        Gizmos.DrawWireSphere (transform.position+centerOfMass, 0.1);
    }

File: tp2/src/CrazyTaxi/Assets/Scripts/Player/CarControleScript.cs
using UnityEngine;
using System.Collections;

public class CarControleScript : MonoBehaviour {		

	public Vector3 centerOfMass = new Vector3(0f,-1.2f,0f);	//Center of mass
	public WheelCollider dataWheel;	//Wheel Collider from which you want to calculate the speed
	public float lowestSteerAtSpeed = 50;	//if lowestSteerAtSpeed < currentSpeed the steer Angle = highSpeedSteerAngel
	public float lowSpeedSteerAngel = 23;	//This could be a high value
	public float highSpeedSteerAngel = 5;	//This shouldn't be a high value (recomended for stability of car)
	public float decellarationSpeed = 40;	//How fast the car will decellarate
	public float maxTorque  = 10;	//Maximum Torque
	public float currentSpeed = 0;		//Current Speed of car
	public float topSpeed = 140;		//Highest speed at which the car can go
	public float maxReverseSpeed = 40; 	//Highest Reverse speed
	public GameObject backLightObject;	//Mesh for reverse light
	public Material idleLightMaterial;	//for idle state
	public Material brakeLightMaterial; 	//Braked state
	public Material reverseLightMaterial;	//Reverse state
//	@HideInInspector
	bool braked = false;	//Brake trigger
	public float maxBrakeTorque = 100; 	//Braking speed
	public Texture2D speedOMeterDial;	//GUI Texture for dial
	public Texture2D speedOMeterPointer;		//GUI Texture for needle
	public int[] gearRatio;		//Shift gear at speed
	//GameObject spark;		//OnCollision Spark
	public GameObject collisionSound;	//OnCollision Sound
	public int minAnglePointer = -90;
	public int maxAnglePointer = 180;
	
	void  Start (){
		rigidbody.centerOfMass=centerOfMass; //Center of mass , for this the car should be pointing on z axis
	}
	
	void  FixedUpdate (){
		HandBrake();
	}
	void  Update (){
		BackLight ();
		EngineSound();
		CalculateSpeed();
	}
	
	//Speed Calculation
	
	void  CalculateSpeed (){
		currentSpeed = 2*22/7*dataWheel.radius*dataWheel.rpm*60/1000;
		currentSpeed = Mathf.Round(currentSpeed);
	}
	
	//Light Control
	
	void  BackLight (){
		if (currentSpeed > 0 && Input.GetAxis("Vertical")<0&&!braked){
			backLightObject.renderer.material = brakeLightMaterial;
		}
		else if (currentSpeed < 0 && Input.GetAxis("Vertical")>0&&!braked){
			backLightObject.renderer.material = brakeLightMaterial;
		}
		else if (currentSpeed < 0 && Input.GetAxis("Vertical")<0&&!braked){
			backLightObject.renderer.material = reverseLightMaterial;
		}
		else if (!braked){
			backLightObject.renderer.material = idleLightMaterial;
		}
	}
	
	//Brake Trigger
	
	void  HandBrake (){
		if (Input.GetButton("Jump")){
			braked = true;
		}
		else{
			braked = false;
		}
	}
	//Engine Sound
	
	void  EngineSound (){
		int i;
		for (i= 0; i < gearRatio.Length; i++){
			if(gearRatio[i]> currentSpeed){
				break;
			}
		}
		float gearMinValue = 0.00f;
		float gearMaxValue = 0.00f;
		if (i == 0){
			gearMinValue = 0;
		}
		else {
			gearMinValue = gearRatio[i-1];
		}
		gearMaxValue = gearRatio[i];
		float enginePitch = ((currentSpeed - gearMinValue)/(gearMaxValue - gearMinValue))+1;
		audio.pitch = enginePitch;
	}
	
	//Speedometer
	
	void  OnGUI (){
		GUI.DrawTexture( new Rect(Screen.width - 300,Screen.height - 300,300,300),speedOMeterDial);
		float speedFactor = currentSpeed / topSpeed;
		float rotationAngle;
		Vector2 pivotPoint;
		if (currentSpeed >= 0){
			rotationAngle = Mathf.Lerp(minAnglePointer,maxAnglePointer,speedFactor);
		}
		else {
			rotationAngle = Mathf.Lerp(minAnglePointer,maxAnglePointer,-speedFactor);
		}
		pivotPoint = new Vector2(Screen.width - 150 ,Screen.height - 150);
		GUIUtility.RotateAroundPivot (rotationAngle, pivotPoint); 
		//GUIUtility.RotateAroundPivot(rotationAngle,Vector2(Screen.width - 150 ,Screen.height - 150));
		GUI.DrawTexture( new Rect(Screen.width - 300,Screen.height - 300,300,300),speedOMeterPointer);
	}
	
	//CollisioN FX
	
	void  OnCollisionEnter ( Collision other  ){
		
		if (other.transform != transform && other.contacts.Length != 0){
			for (int i= 0; i < other.contacts.Length ; i++){
				//Instantiate(spark,other.contacts[i].point,Quaternion.identity);
				GameObject clone = (GameObject)Instantiate(collisionSound,other.contacts[i].point,Quaternion.identity);
				clone.transform.parent = transform;
			}
		}
	}
	
	void  OnDrawGizmos (){
		Gizmos.color = Color.white;
		Gizmos.DrawWireSphere (transform.position+centerOfMass, 0.1f);
	}
	
}